### PR TITLE
"Pad" shorter versions with empty parts when comparing

### LINKED ISSUE #4998: Error comparing versions
### Brief description of your issue

See the discussion -
* https://github.com/microsoft/winget-pkgs/discussions/192959

TL;DR 

```
    RequireLessThan("1-rc", "1"); // Test case pases
    RequireLessThan("1.2-rc", "1.2"); // Test case passes
    RequireLessThan("1.0-rc", "1.0"); // Test case fails
```

### PR DESCRIPTION
**Pull request by Trenly**
- [x] I have signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs).
- [x] This pull request is related to an issue.
  - Resolves #4998 

When versions are created, any empty parts are trimmed off. This means that `22.0.0` is stored as `[22]` in memory. This causes an issue when comparing to a version that had additional parts, but should be sorted lower than the version which is trimmed, such as `22.0.0-rc`.  This is due to the comparator seeing that `[22]` has no more parts to compare after the first iteration, while `[22, 0, 0-rc]` does, causing it to hit the condition where whichever version has more parts is considered to be greater.

This PR updates the logic so that the version comparison effectively contains as many empty parts as are needed to fully complete the comparison. This retains the logic that if a part is not equal to the part it is being compared against, the result of that comparison will be returned but eliminates the need for checking if the versions have the same number of parts since, effectively, they do; it's just that the extra parts needed to complete the comparison are all empty.

Additional tests have been added to cover this corner case.

If possible, I would ask that this also be cherry-picked into any future 1.9 releases (assuming this PR passes review and merges)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/microsoft/winget-cli/pull/5001)

### CODE REVIEW

**Review by florelis — DISMISSED** [review]

File: src/AppInstallerCLITests/Versions.cpp [hunk 1]
```diff
@@ -154,6 +154,14 @@ TEST_CASE("VersionCompare", "[versions]")
     RequireLessThan("0.0.1-beta", "0.0.2-alpha");
     RequireLessThan("13.9.8", "14.1");
 
+    // Ensure that versions with non-digit characters in their parts are sorted correctly
+    RequireLessThan("1-rc", "1");
+    RequireLessThan("1.2-rc", "1.2");
+    RequireLessThan("1.0-rc", "1.0");
```
[the remark this answers is not in the document]
Comment (Trenly) [reply src/AppInstallerCLITests/Versions.cpp:161]: Added!

**Review by florelis — APPROVED** [review]

**Comment by Trenly** [comment]
@florelis - As far as I can tell, the test failures aren’t a result of this change. 

Interestingly, when I run all the tests on the master branch, I get a failure

IMG_6528

-----

The version tests (which should be the only ones impacted by this change) all succeed

IMG_6529

**Review by florelis — APPROVED** [review]
We should update the [comment explaining the sorting](https://github.com/microsoft/winget-cli/blob/bcd8201f8309ec59fa37e870ae33a0983bb5686e/src/AppInstallerSharedLib/Public/AppInstallerVersions.h#L25)

**Review by leighmcculloch — COMMENTED** [review]
🎉 Thanks for finding a fix so quickly.

I had one thought about some test cases to add, but I'm just commenting as an observer and my comments may not be aligned with the focus of the review.

**Comment by Trenly** [comment]
> We should update the [comment explaining the sorting](https://github.com/microsoft/winget-cli/blob/bcd8201f8309ec59fa37e870ae33a0983bb5686e/src/AppInstallerSharedLib/Public/AppInstallerVersions.h#L25)

Agreed and updated

File: src/AppInstallerCLITests/Versions.cpp [hunk 2]
```diff
@@ -154,6 +154,15 @@ TEST_CASE("VersionCompare", "[versions]")
...
     RequireLessThan("13.9.8", "14.1");
 
+    // Ensure that versions with non-digit characters in their parts are sorted correctly
+    RequireLessThan("1-rc", "1");
+    RequireLessThan("1.2-rc", "1.2");
+    RequireLessThan("1.0-rc", "1.0");
+    RequireLessThan("1.0.0-rc", "1");
+    RequireLessThan("22.0.0-rc.1", "22.0.0");
+    RequireLessThan("22.0.0-rc.1", "22.0.0.1");
+    RequireLessThan("22.0.0-rc.1", "22.0.0.1-rc");
```
[the remark this answers is not in the document]
Comment (Trenly) [reply src/AppInstallerCLITests/Versions.cpp:164]: Since I was making changes for Flor's review anyways, added these additional tests

**Review by florelis — APPROVED** [review]

**Comment by Trenly** [comment]
@florelis - Should I raise a cherry pick into 1.9, or has that next patch already been built?

**Comment by florelis** [comment]
> @florelis - Should I raise a cherry pick into 1.9, or has that next patch already been built?

Go ahead and create a PR for the cherry-pick. I'm fine with bringing the change to 1.9, but I'll leave the decision to @denelon / @JohnMcPMS 

We haven't made a build with the latest changes, and most likely won't until after the holiday weekend.
